File: 61-rotate-list/61-rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function (head, k) {
  if (!k || !head || !head.next) return head;

  let [first, second, n] = [head, head, 1];

  while (second.next) {
    n++;
    second = second.next;
  }

  for (let i = 1; i < n - (k % n); i++) first = first.next;

  second.next = head;
  head = first.next;
  first.next = null;

  return head;
};

File: 61-rotate-list/61-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not k or not head or not head.next:
            return head
          
        first, second, n = head, head, 1
        while second.next:
            n += 1
            second = second.next
            
        for i in range(n - k%n - 1):
            first = first.next
            
        second.next = head
        head = first.next
        first.next = None
        
        return head